### ISSUE TITLE
Provide a way to opt-out of discovery time expanding of ITestDataSource tests

### ISSUE THREAD

**Issue by Haplois** [microsoft/testfx#905]
## Description
We recently changed how `ITestDataSource` tests are discovered. We need to provide a way to opt-out from this behavior.

## Proposal
 - Ability to opt-out assembly, type or test case level by using an attribute.
 - Ability to opt-out using a command-line parameter or `.runsettings` setting.

## Steps to reproduce
```csharp
protected static readonly Guid _subscriptionId = Guid.NewGuid();

[DynamicData(nameof(GetData), DynamicDataSourceType.Method)]
[DataTestMethod]
public void MyTest1(Guid subscriptionId)
{
}

private static IEnumerable<object[]> GetData()
{
    yield return new object[]
    {
        _subscriptionId
    };
}
```

## Expected behavior
```
Discovered Tests: 
 - MyTest1
```

## Actual behavior
```
Discovered Tests:
 - MyTest1 (e6a883d7-03b5-4231-a06f-99e5847eb9b0)
```

[AB#1353020](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1353020)

**Comment by patest-dev** [comment]
The opt-out is very(!!) much needed and all this new discovery time behavior should actually be an opt-in.
How could all of these breaking changes be introduced on a minor version??

Even a DynamicData test as simple as this one, is now only being executed 2-times and the `null` case is simply lost

```csharp
private static IEnumerable<object[]> GetData()
{
    List<object[]> ret = new List<object[]>()
    {
        new object[] { null },
        new object[] { "" },
        new object[] { "a" },
    };
    return ret;
}

[TestMethod]
[DynamicData(nameof(GetData), DynamicDataSourceType.Method)]
public void StringTest(string test)
{
    Console.WriteLine(test);
}
```

**Comment by Haplois** [comment]
Support is added by https://github.com/microsoft/testfx/commit/9a621beab927612b357c3f68f97ead58c6282864 to version 2.2.6. Currently we only support assembly level configuration via an attribute. 

`.testsettings` file and class & method level configuration is not possible yet. Depending on feedback, we can add it.

**Comment by abatishchev** [comment]
@Haplois thank you! can you please document this and provide a timeline when the nuget package will be available on nuget.org?

**Comment by abatishchev** [comment]
@Haplois do you face any blockers on the way to release it?

**Comment by abatishchev** [comment]
@Haplois what should be the package versions in order to consume the new functionality?

```xml
<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
<PackageReference Include="MSTest.TestAdapter" Version="2.2.6" />
<PackageReference Include="MSTest.TestFramework" Version="2.2.6" />
```

Like this?

**Comment by Haplois** [comment]
> @Haplois what should be the package versions in order to consume the new functionality?
> 
> ```
> <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
> <PackageReference Include="MSTest.TestAdapter" Version="2.2.6" />
> <PackageReference Include="MSTest.TestFramework" Version="2.2.6" />
> ```
> 
> Like this?

Yes. That's correct. You also need to define [`TestDataSourceDiscoveryAttribute`][1] on assembly level like:

```csharp
[assembly: TestDataSourceDiscovery(TestDataSourceDiscoveryOption.DuringExecution)]
```

[1]: https://github.com/microsoft/testfx/blob/rel/2.2.6/src/TestFramework/MSTest.Core/Attributes/TestDataSourceDiscoveryAttribute.cs#L12

**Comment by ttorkelson** [comment]
> The opt-out is very(!!) much needed and all this new discovery time behavior should actually be an opt-in. How could all of these breaking changes be introduced on a minor version??
> 
> Even a DynamicData test as simple as this one, is now only being executed 2-times and the `null` case is simply lost
> 
> ```cs
> private static IEnumerable<object[]> GetData()
> {
>     List<object[]> ret = new List<object[]>()
>     {
>         new object[] { null },
>         new object[] { "" },
>         new object[] { "a" },
>     };
>     return ret;
> }
> 
> [TestMethod]
> [DynamicData(nameof(GetData), DynamicDataSourceType.Method)]
> public void StringTest(string test)
> {
>     Console.WriteLine(test);
> }
> ```

I totally agree that this should be an opt-in rather than an opt-out. These changes caused quite a bit of headache in my org

**Comment by abatishchev** [comment]
> These changes caused quite a bit of headache in my org

Preach!!

**Comment by gdir22** [comment]
> I totally agree that this should be an opt-in rather than an opt-out. These changes caused quite a bit of headache in my org

I have a library that perfoms mathematical operations. I only investigated in this problem because I had some tests that where passing up to 2.2.3, but started to fail whenever I tried to update to 2.2.4 - 2.2.10. Only during debug I found out that a lot of the still passing tests in 2.2.4 - 2.2.10 only passed, because my immutable numerical structs where replaced by zeros. Instead of testing the real test cases (as in 2.2.3), I was actually mostly testing 0.0 == 0.0 in 2.2.4 - 2.2.10 and the tests where wrongfully passing.

I tried to look in the documentation, but I actually found the solution only by searching through the issues. I agree with @ttorkelson that this is a major and breaking change in behavior that should not be introduced in a minor version. There should be at least a compiler warning that there is a risk that immutable, not serializable structs as test parameters are not handled correctly when TestDataSourceDiscovery is not set to DuringExecution.
